### PR TITLE
Authentication Provider Best Practices: Centralized Login

### PR DESCRIPTION
**Pull request by kmaida**
Needs design / assets / SEO / title finalization.

### CODE REVIEW

File: _posts/authentication-provider-best-practices-centralized-login.markdown [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+
+### What is Embedded Login?
+
+**Embedded login** refers to a method of authentication wherein credentials are entered via an experience that is _embedded_ on a web app's domain or in a WebView (in the case of native apps). Credentials are then sent to the authentication provider for login. In a web app, this is a _cross-origin_ request. Embedded logins present a range of potential security and implementation challenges that cause issues for developers and users; as a matter of fact, [Google no longer supports an embedded approach when implementing OAuth](https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html).
```
Comment (unicodeveloper) [inline _posts/authentication-provider-best-practices-centralized-login.markdown]: Shouldn't we reference our own article https://auth0.com/blog/google-blocks-oauth-requests-from-embedded-browsers/ here ?

File: _posts/authentication-provider-best-practices-centralized-login.markdown [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+{% include tweet_quote.html quote_text="Centralized login provides security, SSO, maintainability, best practice native app implementation, and more." %}
+
+### Security
+
+Centralized login is more secure than embedded login. Authentication takes place over the same domain, eliminating cross-origin requests. Cross-origin authentication is inherently more dangerous. Collecting user credentials in an application served from one origin and then sending them to another origin can present certain security vulnerabilities. [Phishing attacks](https://auth0.com/blog/all-you-need-to-know-about-the-google-docs-phishing-attack/) are more likely, as are [man-in-the-middle attacks](https://auth0.com/docs/security/common-threats#man-in-the-middle-mitm-attacks). Centralized login does not send information between origins, thereby negating cross-origin concerns.
+
+Embedded user agents are unsafe for third parties, including the authorization server itself. If an embedded login is used, the app has access to both the authorization grant and the user's authentication credentials. As a consequence, this data is left vulnerable to recording or malicious use.
```
Comment (unicodeveloper) [inline _posts/authentication-provider-best-practices-centralized-login.markdown]: > If an embedded login is used, the app has access to both the authorization grant and the user's authentication credentials. As a consequence, this data is left vulnerable to recording or malicious use.

By default, an app that utilizes a custom login feature has access to a user's authentication credentials. Are you saying here that the developers of the app might record or use it for something malicious? Something is not clear here. Might need rephrasing for better clarification.

File: _posts/authentication-provider-best-practices-centralized-login.markdown [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+### Single Sign-On
+
+Centralized login orchestrates [single sign-on](https://auth0.com/docs/sso/current) (SSO) between multiple apps while providing cookies from the same origin. Once a user has logged in using a centralized login, a cookie is created and stored. Any future calls to the authentication provider's authorization endpoint will then check the cookie. If the user has already signed on, the login page will not be shown again and the user will be logged in via SSO. On the other hand, embedded user agents don't share authentication state, meaning that they cannot be used for SSO.
+
+### Easier to Implement and Maintain
+
+Centralized login is easier to implement as well as maintain for app developers. Cross-origin authentication is inherently more dangerous, but centralized login mitigates this risk entirely. A centralized login page is already fully implemented, negating the need for the developer to build out the login UI if a custom UI is not required. The authorization server providing the centralized login page can also ensure a consistent and secure experience across all apps that utilize it.
```
Comment (unicodeveloper) [inline _posts/authentication-provider-best-practices-centralized-login.markdown]: Easier to implement? Is it really easier to implement compared to the embedded login? 🤔

Comment (unicodeveloper) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: For maintenance, I agree it is easier.

File: _posts/authentication-provider-best-practices-centralized-login.markdown [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+$ NODE_ENV=dev node server
+```
+
+You should now be able to run the app and authenticate. You should then also be able to post to the message board, sending your access token to the provided Node API.
+
+We now have a real-world MEAN stack application with Auth0 centralized login and passwordless authentication! 
+
+## Conclusion
+
+Centralized login is the most secure and maintainable standards-based approach to logging in with an authentication provider. Unlike embedded login, it is safer from cross-origin attack vectors and poses no danger to the authorization server. Centralized login is [best current practice for native mobile apps](https://auth0.com/blog/oauth-2-best-practices-for-native-apps/), and OAuth providers like [Google no longer support embedded login strategies](https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html).
```
Comment (unicodeveloper) [inline _posts/authentication-provider-best-practices-centralized-login.markdown]: > Centralized login is best current practice for native mobile apps

I think you omitted `the` before `best current....`

Comment (unicodeveloper) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: Again, I think we should link to our article instead https://auth0.com/blog/google-blocks-oauth-requests-from-embedded-browsers/

File: _posts/authentication-provider-best-practices-centralized-login.markdown [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+- authentication
+- best-practices
+- security
+- centralized-login
+- angular
+- mean
+- node
+- passwordless
+related:
+- 2017-10-19-oauth-2-best-practices-for-native-apps
```
Comment (unicodeveloper) [inline _posts/authentication-provider-best-practices-centralized-login.markdown]: This link https://auth0.com/blog/google-blocks-oauth-requests-from-embedded-browsers/ can be part of the related links here.

Comment (kmaida) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: Great! I didn't know that this article existed. :) 👍

File: _posts/authentication-provider-best-practices-centralized-login.markdown (hunk 1 above)
Comment (kmaida) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: 👍

File: _posts/authentication-provider-best-practices-centralized-login.markdown (hunk 4 above)
Comment (kmaida) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: 👍

File: _posts/authentication-provider-best-practices-centralized-login.markdown (hunk 2 above)
Comment (kmaida) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: Yes, basically, this is referring to this section of the RFC:


> In typical web-view-based implementations of embedded user-agents,
   the host application can record every keystroke entered in the login
   form to capture usernames and passwords, automatically submit forms
   to bypass user consent, and copy session cookies and use them to
   perform authenticated actions as the user.

> Even when used by trusted apps belonging to the same party as the
   authorization server, embedded user-agents violate the principle of
   least privilege by having access to more powerful credentials than
   they need, potentially increasing the attack surface.

 > Encouraging users to enter credentials in an embedded user-agent
   without the usual address bar and visible certificate validation
   features that browsers have makes it impossible for the user to know
   if they are signing in to the legitimate site; even when they are, it
   trains them that it's OK to enter credentials without validating the
   site first.

I'll try to come up with a better way to paraphrase this.

File: _posts/authentication-provider-best-practices-centralized-login.markdown (hunk 3 above)
Comment (kmaida) [reply _posts/authentication-provider-best-practices-centralized-login.markdown]: The ease of implementation is covered in this paragraph regarding centralized login mitigating cross-origin risk and thus meaning the developer does not need to implement additional CORS security or a login UI (in a general sense, not necessarily specifically referring to Auth0's embedded Lock).

I'll try to clarify this further.

File: _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+![centralized vs. embedded login](https://cdn.auth0.com/blog/centralized-login/centralized-embedded-diagram.jpg)
+
+1. **Centralized Login** - When clicking a button or link to authenticate in both the browser and native app, the centralized login URL at [accounts.google.com](https://accounts.google.com) is loaded.
+2. **Embedded Login** - When clicking a button or link to authenticate in the browser, an overlay modal is opened on the same domain, prompting the user to log in. The mobile app also displays input fields within the app.
+
+## A Tale of Two Companies
+
+Let's begin with two hypothetical timelines from the perspective of tech teams at companies with imaginary products. These examples can help us visualize and relate to the challenges presented when implementing authentication in a way that doesn't afford enough flexibility.
+
```
Comment (aaguiarz) [inline _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown]: I don't understand the comparison. Is it comparing a company not using Auth0 and rolling their own authentication vs a company using Auth0, or a company using embedded login vs centralized login. 

It seems to be comparing the first two, but it should be comparing the second ones.

It seems that the main difference is that there's SSO and that if Google acquires you life is easier ;)

File: _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+### What is Embedded Login?
+
+**Embedded login** refers to a method of authentication wherein credentials are entered via an experience that is _embedded_ on a web app's domain with text inputs, or in the case of native apps, with inputs or an embedded WebView. Credentials are then sent to the authentication provider for login. In a web app, this is a _cross-origin_ request. Embedded logins present a range of potential security and implementation challenges that cause issues for developers and users; as a matter of fact, [Google no longer supports an embedded approach when implementing OAuth](https://auth0.com/blog/google-blocks-oauth-requests-from-embedded-browsers/).
```
Comment (aaguiarz) [inline _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown]: Is an "Embedded WebView" embedded login? I'd say it's a bad practice for implementing centralized login, but it's not embedded.

File: _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown (hunk 6 above)
Comment (kmaida) [reply _posts/2017-11-02-authentication-provider-best-practices-centralized-login.markdown]: Yes, this section is actually going to be rewritten. :) @woloski and I agree the example comparison does not make it very clear what the difference and advantages are. I'm going to turn it into more of a hypothetical case study instead and this should be better.

**Pull request merged**